Sources/Filters/General/ContourTriangulator/helper.js

```diff
@@ -0,0 +1,2290 @@
+ * Shorter edges will be preferred over long edges.
+ *
+ * @param {Number[][]} polys
+ * @param {Number[]} incompletePolys
+ * @param {vtkPoints} points
+ * @param {Vector3} normal
+ */
+export function vtkCCSJoinLooseEnds(polys, incompletePolys, points, normal) {
+  // Relative tolerance for checking whether an edge is on the hull
+  const tol = CCS_POLYGON_TOLERANCE;
+
+  // A list of polys to remove when everything is done
+  const removePolys = [];
+
+  const p1 = [];
+  const p2 = [];
+  let poly1;
+  let poly2;
+  let pt1;
+  let pt2;
+  let dMin;
+  let iMin;
+  let v;
+  let d;
+
+  let n = incompletePolys.length;
+  while (n !== 0) {
+    poly1 = polys[incompletePolys[n - 1]];
+    pt1 = poly1[poly1.length - 1];
+    points.getPoint(pt1, p1);
+
+    dMin = Number.MAX_VALUE;
+    iMin = 0;
+
+    for (let i = 0; i < n; i++) {
+      poly2 = polys[incompletePolys[i]];
+      pt2 = poly2[0];
+      points.getPoint(pt2, p2);
+
+      // The next few steps verify that edge [p1, p2] is on the hull
+      v = [p2[0] - p1[0], p2[1] - p1[1], p2[2] - p1[2]];
+      d = vtkMath.normalize(v);
+      if (d !== 0) {
+        v[0] /= d;
+        v[1] /= d;
+        v[2] /= d;
+      }
+
+      // Compute the midpoint of the edge
+      const pm = [
+        0.5 * (p1[0] + p2[0]),
+        0.5 * (p1[1] + p2[1]),
+        0.5 * (p1[2] + p2[2]),
+      ];
+
+      // Create a plane equation
+      const pc = [];
+      vtkMath.cross(normal, v, pc);
+      pc[3] = -vtkMath.dot(pc, pm);
+
+      // Check that all points are inside the plane.  If they aren't, then
+      // the edge is not on the hull of the pointset.
+      let badPoint = false;
+      const m = polys.length;
+      const p = [];
+      for (let j = 0; j < m && !badPoint; j++) {
+        const poly = polys[j];
+        const npts = poly.length;
+        for (let k = 0; k < npts; k++) {
+          const ptId = poly[k];
+          if (ptId !== pt1 && ptId !== pt2) {
+            points.getPoint(ptId, p);
+            const val = p[0] * pc[0] + p[1] * pc[1] + p[2] * pc[2] + pc[3];
+            const r2 = vtkMath.distance2BetweenPoints(p, pm);
+
+            // Check distance from plane against the tolerance
+            if (val < 0 && val * val > tol * tol * r2) {
+              badPoint = true;
+              break;
+            }
+          }
+        }
+
+        // If no bad points, then this edge is a candidate
+        if (!badPoint && d < dMin) {
+          dMin = d;
+          iMin = i;
+        }
+      }
+    }
+
+    // If a match was found, append the polys
+    if (dMin < Number.MAX_VALUE) {
+      // Did the poly match with itself?
+      if (iMin === n - 1) {
+        // Mark the poly as closed
+        incompletePolys.pop();
+      } else {
+        const id2 = incompletePolys[iMin];
+
+        // Combine the polys
+        // TODO: Check
+        // poly1.insert(poly1.end(), polys[id2].begin(), polys[id2].end());
+        for (let i = 1; i < polys[id2].length; i++) {
```

Comment:
why `let i = 1` ?

Reply:
Because the endpoint of poly1 is equal to the startpoint of poly2.

E.g. combining [0,1,2] and [2,3,4] should give [0,1,2,3,4].

Reply:
but that is not what the c++ code does, does it ?

Reply:
I did some testing and starting from `i=1` reduced the number of polys in the end result, while still being a correct representation of the clipped PolyData.

Clipped with three planes at Origin [0, 0, 0] and normals [1, 0, 0], [0, 1, 0], [0, 0, 1].

(# of polys with `poly1.push(...polys[id2])`) -> (# of polys with for-loop starting from `i=1`)
vtkCubeSource: 18 -> 12
vtkConeSource: 11 -> 10
vtkSphereSource (thetaRes: 20, phiRes: 11): 74 -> 73

I haven't tested it with a more complex polydata yet, but I still thought it may be worth sharing.

Reply:
fair enough. but that should go in a separate commit (same MR).

btw, can you also create a separate (1rst) commit for the changes in:
 - Common/Core/CellArray, 
 - Common/Core/DataArray, 
 - Common/Core/Points
 - Common/DataModel/DataSetAttributes
 and make me a co-author of that commit.